### PR TITLE
Session tracking

### LINKED ISSUE #75: PANIC: bugsnag.Notify panics on nil error
### Expected behaviour
It should not panic/crash. Instead, it should create an event with `message` empty string or it should gracefully omit the event(i'm not sure)

### Observed behaviour
When I tried to create a handled error event, I mistakenly sent `nil error` to `bugsnag.Notify`, on the newEvent creation it tried to read a message from the nil error. It panicked on the server.

### Steps to reproduce
Pass `nil` error into Notify().
```
var err error
....
bugsnag.Notify(err, bugsnag.MetaData{
        "ErrTrace": {
		"AddInfo":    "Some trace",
	},
})
```
```
        /home/ubuntu/goapps/src/code.foobar.in/root/foobar-backend/errorsx/errorsx.go:90 +0x5
code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go/errors.(*Error).Error(0xc421ab8140, 0x1180fcb, 0x11)
        /home/ubuntu/goapps/src/code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go/errors/error.go:84 +0x33
code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go.newEvent(0xc421ab8140, 0xc422185e68, 0x1, 0x1, 0x1e5e960, 0x0, 0x1e86
ae0)
        /home/ubuntu/goapps/src/code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go/event.go:113 +0x17c
code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go.(*Notifier).NotifySync(0x1e5e960, 0x1e86ea0, 0xc421ab8140, 0x0, 0xc42
2185e68, 0x1, 0x1, 0xc422185d98, 0x41bad5)
        /home/ubuntu/goapps/src/code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go/notifier.go:46 +0xa5
code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go.(*Notifier).Notify(0x1e5e960, 0x1e86ea0, 0xc421ab8140, 0xc422185e68, 
0x1, 0x1, 0xc4222a0e30, 0xc422185ea8)
        /home/ubuntu/goapps/src/code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go/notifier.go:38 +0x72
code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go.Notify(0x1e86ae0, 0x0, 0xc422185e68, 0x1, 0x1, 0x3, 0xc4221b0140)
        /home/ubuntu/goapps/src/code.foobar.in/root/foobar-backend/vendor/github.com/bugsnag/bugsnag-go/bugsnag.go:42 +0x9e
code.foobar.in/root/foobar-backend/api.getAndSaveGithubLangs(0xc42131a180, 0x24, 0xc420e89010, 0xc, 0xc420c5b950, 0x28, 0xc42155f280, 0x4, 0x
4)
```
`errors.New(err, 1)`  returning nil.

### Version
1.3.1

Please let me know if it a valid issue.

### PR DESCRIPTION
**Pull request by kinbiko**
## Goal

Add session tracking 🎉 

Also made it easier to automatically add request data to an error/panic by passing in a `context.Context` from the `*http.Request` (or `c.Args["context"]` in the case of Revel).

This PR is intended to stay up for a while to be reviewed whenever time allows. Comments raised on this PR are likely going to be addressed on other branches that will be merged into this (`session-tracking`) branch.

## Changeset

This is not intended to be a breaking release, however this work does increase, and to a certain extent introduce, version requirements for the Go langauge, and/or the various frameworks.

In particular, Go 1.7 is required as of this change. We're also restricting versions of Revel to be between 0.18.0 and 0.20.0, which implicitly means that Go 1.7 is no longer supported as Revel requires 1.8 as of 0.18.0.

- [x] Session delivery logic #68 
- [x] Session tracker implementation #69 
- [x] Migrate payload versions from 2 to 4 #73
- [x] Session tracking for each framework #74
- [x] Maze runner #89 

### Added

New packages:

```
bugsnag-go/
	sessions/
	headers/
	device/
```

The code relating to notifying about errors/panics is still in the root `bugsnag-go` package, to avoid breaking backwards compatibility with the exposed structs. Ideally there'd be a separate package for this logic, but that'll have to come in a breaking release.

### Removed

Removed and replaced the old framework integration tests as they were inconsistent and difficult to work with. All frameworks can easily be used in tests apart from Revel. Revel is largely tested manually and through Maze tests.

### Changed

Discovered quite a few bugs when introducing new tests that were fixed.
These will be outlined in the change log when finished, or can be identified from the commits on this PR.

## Tests

Considerable addition of tests, a large number of which spin up a test server to further add confidence in the payloads that are being sent.

Got rid of the existing integration tests for the various frameworks as they were very difficult to get running (I couldn't get them to pass) and didn't die cleanly. Instead, added new integration tests for each framework, apart from Revel, which is being extensively tested with Maze.

### Linked issues

Fixes #75 - as a side effect.

Related to #44 - this issue requires more thought in an area unrelated to these changes, and should be revisited in the near future. These tests fail due to excessively precise tests that do not take into account changes in the Go runtime in versions newer than when this package was originally created, but the functionality still works as expected. The tests for this package are currently ignored to more accurately reflect the confidence we have in the library. 

## Review

For the submitter, initial self-review:

- [x] Commented on code changes inline explain the reasoning behind the approach
- [x] Reviewed the test cases added for completeness and possible points for discussion
- [x] A changelog entry was added for the goal of this pull request
- [x] Check the scope of the changeset - is everything in the diff required for the pull request?
- This pull request is ready for:
  - [x] Initial review of the intended approach, not yet feature complete
  - [x] Structural review of the classes, functions, and properties modified
  - [x] Final review

For the pull request reviewer(s), this changeset has been reviewed for:

- [ ] Consistency across platforms for structures or concepts added or modified
- [ ] Consistency between the changeset and the goal stated above
- [ ] Internal consistency with the rest of the library - is there any overlap between existing interfaces and any which have been added?
- [ ] Usage friction - is the proposed change in usage cumbersome or complicated?
- [ ] Performance and complexity - are there any cases of unexpected O(n^3) when iterating, recursing, flat mapping, etc?
- [ ] Concurrency concerns - if components are accessed asynchronously, what issues will arise
- [ ] Thoroughness of added tests and any missing edge cases
- [ ] Idiomatic use of the language

### CODE REVIEW

File: event.go [hunk 1]
```diff
@@ -98,35 +99,42 @@ type Event struct {
...
+
+		case error, errors.Error:
+			err = errors.New(datum.(error), 4)
+			event.Error = err
+			event.ErrorClass = err.TypeName()
+			event.Message = err.Error()
+			event.Stacktrace = make([]stackFrame, len(err.StackFrames()))
+
+		case bool:
+			sync = bool(datum)
```
Comment (CodeHex) [inline event.go]: I don't think this will work for the defaultPanicHandler. It currently passes in a config object to make it synchronous https://github.com/bugsnag/bugsnag-go/blob/master/panicwrap.go#L25, which would get ignored. Also any settings from the original config is ignored as it defaults to false.

What you could do is change this line to `config = config.merge(Configuration{Synchronous: bool(datum)})` and go back to using the config.Synchronous value directly. This should pick up it being set directly via a bool, or in a config object.

File: notifier.go [hunk 2]
```diff
@@ -33,23 +33,34 @@ func New(rawData ...interface{}) *Notifier {
 // Notify sends an error to Bugsnag. Any rawData you pass here will be sent to
 // Bugsnag after being converted to JSON. e.g. bugsnag.SeverityError, bugsnag.Context,
 // or bugsnag.MetaData.
-func (notifier *Notifier) Notify(err error, rawData ...interface{}) (e error) {
-	config := notifier.Config
-	return notifier.NotifySync(err, config.Synchronous, rawData...)
+func (notifier *Notifier) Notify(rawData ...interface{}) (e error) {
+	return notifier.NotifySync(append(rawData, notifier.Config.Synchronous)...)
```
Comment (CodeHex) [inline notifier.go]: The order of the raw data variables have changed, leading to a different precedence for deciding if the report should be sent synchronously. Before it would use the config.Synchronous first, but then it could be overwritten by a config object or boolean in the raw data. Now, it doesn't matter what the raw data is, it will always use the config.Synchronous variable since its the last parameter.

Comment (kinbiko) [reply notifier.go]: Note to self: fix this by changing
```go
NotifySync(append(rawData, notifier.Config.Synchronous)...)
```
into
```go
NotifySync(append(arrayContainingConfigSynchronous, allElementsOfRawData)...)
```

File: bugsnag.go [hunk 3]
```diff
@@ -57,21 +88,32 @@ func AutoNotify(rawData ...interface{}) {
...
-		defaultNotifier.NotifySync(errors.New(err, 2), true, rawData...)
+		defaultNotifier.NotifySync(append(rawData, errors.New(err, 2), true)...)
 		panic(err)
 	}
 }
 
 // Recover logs a panic on a goroutine and then recovers.
+// Although it's not strictly enforced, it's highly recommended to pass a
+// context.Context object that has at one-point been returned from
+// bugsnag.StartSession. Doing so ensures your stability score remains accurate,
```
Comment (CodeHex) [inline bugsnag.go]: In this case, the stability score is unaffected because Recover will report crashes as handled and won't contribute to the stability score. We may need to provide another recover method or change the example to include how to set it up to affect the stability score.

File: sessions/tracker.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		sessions:       []*Session{},
+		config:         config,
+		publisher:      &publisher,
+	}
+	go st.processSessions()
+	return &st
+}
+
+func (s *sessionTracker) GetSession(ctx context.Context) *Session {
+	return ctx.Value(contextSessionKey).(*Session)
```
Comment (CodeHex) [inline sessions/tracker.go]: I think this will crash if someone forgets to add the session to the context thats passed into the autonotify

File: sessions/tracker.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+
+func (s *sessionTracker) interval() time.Duration {
+	s.config.mutex.Lock()
+	defer s.config.mutex.Unlock()
+	return s.config.PublishInterval
+}
+
+func (s *sessionTracker) processSessions() {
+	tic := time.Tick(s.interval())
+	shutdown := shutdownSignals()
```
Comment (CodeHex) [inline sessions/tracker.go]: Is it possible to flush sessions when the defaultPanicHandler kicks in? Currently, if a unhandled panic occurs which is reported by the defaultPanicHandler, any unflushed sessions are lost

File: revel/request_extractor.go [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+		Referer:    req.Header.Get("referer"),
+	}
+}
+
+// extractHeader is a best-guess workaround for pulling out as many headers as possible in a Revel request.
+// Extracts all the standard HTTP headers, and some non-standard, but common headers.
+func extractHeaders(h headerExposer, paramsFilters []string) map[string]string {
+	m := make(map[string]string)
+	// Standard headers
+	addHeader(paramsFilters, m, h, "Accept")
```
Comment (fractalwrench) [inline revel/request_extractor.go]: Would it be possible to add the header keys to some sort of collection then iterating over it, to avoid multiple calls to `addHeader`?

File: bugsnag.go [hunk 7]
```diff
@@ -127,24 +194,48 @@ func init() {
...
 	})
 
 	hostname, err := os.Hostname()
 	if err == nil {
 		Config.Hostname = hostname
 	}
 }
+
+func startSessionTracking() {
+	if sessionTracker == nil {
```
Comment (fractalwrench) [inline bugsnag.go]: Is there a potential for multiple sessionTrackers to be instantiated here if this method is called concurrently?

File: sessions/payload.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+
+// makeSessionPayload creates a sessionPayload based off of the given sessions and config
+func makeSessionPayload(sessions []*Session, config *SessionTrackingConfiguration) *sessionPayload {
+	releaseStage := config.ReleaseStage
+	if releaseStage == "" {
+		releaseStage = "production"
+	}
+	hostname := config.Hostname
+	if hostname == "" {
+		hostname, _ = os.Hostname() //Ignore the hostname if this call errors
```
Comment (fractalwrench) [inline sessions/payload.go]: have we considered caching this value? See https://thoeni.io/post/macos-sierra-java/

File: sessions/payload.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+			ReleaseStage: releaseStage,
+		},
+		Device: devicePayload{
+			OsName:   runtime.GOOS,
+			Hostname: hostname,
+		},
+		SessionCounts: []sessionCountsPayload{
+			{
+				//This timestamp assumes that we're sending these off once a minute
+				StartedAt:       sessions[0].StartedAt.UTC().Format(time.RFC3339),
```
Comment (fractalwrench) [inline sessions/payload.go]: is this array guaranteed to be non-empty?

File: CONTRIBUTING.md [hunk 10]
```diff
@@ -71,11 +71,25 @@ If you are a project maintainer, you can build and release a new version of
...
+
+1. Install the [Revel CLI tool](https://revel.github.io/tutorial/gettingstarted.html)
+1. Install [json-server](https://github.com/typicode/json-server): `npm install -g json-server`.
+1. `cd` into the `examples/revelapp/` directory
+1. Run the `start-test-servers.sh` script
+1. Run `revel run`
+1. Visit `localhost:4321/handled` and verify that `payloads.json` now contains a report payload containing a handled payload, including a `session` sub-object containing 0 unhandled and 1 handled.
+1. Visit `localhost:4321/unhandled` and verify that `payloads.json` now contains an unhandled payload, including a `session` sub-object containing 1 unhandled and 0 handled.
+1. Let the application run for at least one minute and verify that sessions are being populated under the "sessions" array in `payloads.json`
+1. Make sure you do not commit the `payload.json` file.
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply CONTRIBUTING.md]: The file needs to be present, and as far as I know, there's no way of ignoring a previously added file that we don't wish to delete first; at least not ignoring it so that it applies to the remote repository as well.

If you know of a way to do this, let me know.

File: examples/revelapp/README.md [hunk 11]
```diff
@@ -1,11 +1,43 @@
...
-## Configuration
+A high-productivity web framework for the [Go language](http://www.golang.org/).
 
-1. Install revel:
 
-       go get -u github.com/revel/cmd/revel
+### Start the web server:
 
-2. Insert API key in `conf/app.conf`
+   revel run myapp
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply examples/revelapp/README.md]: This is a default revel README for a recently generated app that accidentally overwrote the README previously in this location.

I'll update this README to be specific to our example app instead.

File: examples/revelapp/README.md [hunk 12]
```diff
@@ -1,11 +1,43 @@
...
+
+### Go to http://localhost:9000/ and you'll see:
+
+    "It works"
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply examples/revelapp/README.md]: This is a default revel README for a recently generated app that accidentally overwrote the README previously in this location.

I'll update this README to be specific to our example app instead.

File: examples/revelapp/conf/app.conf [hunk 13]
```diff
@@ -1,49 +1,213 @@
...
+
+# Sets `revel.AppName` for use in-app.
+# Example:
+#   `if revel.AppName {...}`
+app.name = revelapp
+
+# Bugsnag config
+bugsnag.apikey=066f5ad3590596f9aa8d601ea89af845
+bugsnag.endpoints.notify=http://localhost:1234/reports
+bugsnag.endpoints.sessions=http://localhost:1234/sessions
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply examples/revelapp/conf/app.conf]: You're absolutely right.
This example is intended to be updated before merging into master. Its primary intent (right now, prior to maze tests) is to serve for manual testing, which these endpoints are set up for. The contribution guidelines outline how this works.

File: configuration.go [hunk 14]
```diff
@@ -75,16 +103,17 @@ type Configuration struct {
 	// Whether bugsnag should notify synchronously. This defaults to false which
 	// causes bugsnag-go to spawn a new goroutine for each notification.
 	Synchronous bool
+	// Whether the notifier should send all sessions recorded so far to Bugsnag
+	// when repanicking to ensure that no session information is lost in a
+	// fatal crash.
+	flushSessionsOnRepanic bool
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply configuration.go]: Booleans are really awkward configuration options in this notifier. Instead of addressing this issue (non-trivial, probably breaking changes required) or exposing another hard-to-understand configuration option, I've exposed a function that sets this property, and have it set to true by default - like you suggested. 

However, we don't always want to do this because this could lead to a very high amount of network traffic for services that have a lot of panics that are handled by a framework. We only want to flush in the case that we're pretty sure the application will die. Not when we just see a panic, because there might be a framework panic handler to catch it. We disable this setting in all of our framework integrations.

File: payload_test.go [hunk 15]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/bugsnag/bugsnag-go/sessions"
+)
+
+const expSmall = `{"apiKey":"","events":[{"app":{"releaseStage":""},"device":{},"exceptions":[{"errorClass":"","message":"","stacktrace":null}],"metaData":{},"payloadVersion":"4","severity":"","unhandled":false}],"notifier":{"name":"Bugsnag Go","url":"https://github.com/bugsnag/bugsnag-go","version":"1.4.0"}}`
+
+// The large payload has a timestamp in it which makes it awkward to assert against.
+// Instead, assert that the timestamp property exist, along with the rest of the expected payload
+const expLargePre = `{"apiKey":"166f5ad3590596f9aa8d601ea89af845","events":[{"app":{"releaseStage":"mega-production","type":"gin","version":"1.5.2"},"context":"/api/v2/albums","device":{"hostname":"super.duper.site"},"exceptions":[{"errorClass":"error class","message":"error message goes here","stacktrace":[{"method":"doA","file":"a.go","lineNumber":65},{"method":"fetchB","file":"b.go","lineNumber":99,"inProject":true},{"method":"incrementI","file":"i.go","lineNumber":651}]}],"groupingHash":"custom grouping hash","metaData":{"custom tab":{"my key":"my value"}},"payloadVersion":"4","session":{"startedAt":"`
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply payload_test.go]: I did consider doing that. Small changes in these tests would mean that we also had to change fixture files, which should really live in a subdirectory somewhere to avoid polluting this package. But since it's only used by one test file I decided against doing that. It's easier to see where related code lives this way.

File: sessions/payload.go (hunk 9 above)
Comment (kinbiko) [reply sessions/payload.go]: Yes, however the check for that is somewhat distant from this line. I can only assume we try and be pretty defensive about this stuff, so I'll add another check a bit more locally.

File: bugsnag.go (hunk 7 above)
Comment (kinbiko) [reply bugsnag.go]: Yeah, good catch. Ensured that startSessionTracking only gets called once.

File: event.go [hunk 16]
```diff
@@ -175,3 +191,34 @@ func newEvent(err *errors.Error, rawData []interface{}, notifier *Notifier) (*Ev
...
+	}
+
+	event.Request = extractRequestInfoFromReq(req)
+
+	if event.Context == "" {
+		event.Context = req.URL.Path
+	}
+
+	// Default user.id to IP so that the count of users affected works.
+	if event.User == nil {
```
[the remark this answers is not in the document]
Comment (kinbiko) [reply event.go]: This is already documented, it was just moved as part of this work.

https://docs.bugsnag.com/platforms/go/net-http/automatically-captured-data/#user-information
